### DISCUSSION TITLE
[Help] Mini.align - How to ignore delimiter if it's inside double quotes

### DISCUSSION THREAD

**Discussion by icyrainz** [nvim-mini/mini.nvim#361] [Q&A] [answered]
As title, google sheet CSV format would put each column data as is if there is no comma inside and will wrap the column data inside double quote to differentiate with legitimate delimiter.

I have gone through the help file but could not a way to do this. Not sure if the plugin supports this capability or not.
```
Col1,Col2,Col2
Data1,Data2,Data3
"Data1WithComma,OtherSide",Data2,Data3
```
gAip,
-> The first column of the second entry should be `"Data1WithComma,OtherSide"` and not `"Data1WithComma`

**Comment by echasnovski** [comment]
**Edit**: Somewhat ignore this answer. There **is** a better built-in interactive way.

This question is relevant to me too, as I do quite frequently work with .csv files. But I don't think I found a way to organically fit this type of separator ignore into **interactive** mode of 'mini.align'.

There are two approaches here I can recommend:
- Manually filter out separators which you don't want to include. For your example it would be `gAip,` followed by `f` (built-in "filter" modifier) and typing `not (row == 3 and col == 1)` and `<CR>` (`<Enter>`). This will add `filter` pre-justify step which will exclude one separator from aligning.
- Write custom modifier which splits by commas not inside quotes. I had time and was curious and here is what I ended up with:
    ```lua
    local split_csv_row = function(row_string)
      -- Compute "forbidden" spans
      local forbidden = {}
      local add_to_forbidden = function(l, r) table.insert(forbidden, { from = l + 1, to = r }) end

      row_string:gsub("()%b''()", add_to_forbidden)
      row_string:gsub('()%b""()', add_to_forbidden)

      -- Compute all comma positions which are not inside forbidden spans
      local commas = {}
      row_string:gsub('%,()', function(r)
        for _, span in ipairs(forbidden) do
          if span.from <= r and r <= span.to then return end
        end
        table.insert(commas, r - 1)
      end)

      -- Split
      local n = #commas
      if n == 0 then return { row_string } end

      local res = { row_string:sub(1, commas[1]) }
      for i = 1, n - 1 do
        table.insert(res, row_string:sub(commas[i] + 1, commas[i + 1]))
      end
      table.insert(res, row_string:sub(commas[n] + 1))

      return res
    end

    local split_csv_step = require('mini.align').new_step(
      'csv-row',
      function(strings, _) return vim.tbl_map(split_csv_row, strings) end
    )

    require('mini.align').setup({
      modifiers = {
        c = function(steps, opts)
          steps.split = split_csv_step
          opts.split_pattern = '"csv-row"'
        end,
      },
    })
    ```
    Now you can `gAip` followed by `c` and it should split by commas ignoring spans inside `''` and `""`.
    Please note, that this is not thoroughly tested and might need extra tweaks, but seems to work in the cases I tried.

**Comment by echasnovski** [comment, marked as answer]
Apparently, I forgot how my modules work :disappointed: 

There is an `i` modifier (for `ignore`) that does just that: adds matches inside `""` to be ignored. So you can `gAip` followed by `,` to set separator **and** followed by `i` to add "ignore" pre-split step.

If you want to ignore inside both `''` and `""`, you can customize `i` modifier like so:

```lua
local align = require('mini.align')
align.setup({
  modifiers = {
    i = function(steps, _) table.insert(steps.pre_split, align.gen_step.ignore_split({ '".-"', "'.-'" })) end,
  },
})
```

**Reply by icyrainz** [reply]
Thanks for the quick response. Works like a charm.
